tests/test_infer.py

```diff
@@ -1064,3 +1065,62 @@ def g(c):
         else:
             return f3
     return g(x < 42)(x)
+
+
+@infer(type=[(i64, i64)],
+       shape=[({'type': ai64, 'shape': (6, 13)}, (6, 13))])
+def test_identity(x):
+    return identity(x)
+
+
+@infer(type=[(B, B, B),
+             (i64, B, InferenceError),
+             (B, i64, InferenceError)])
+def test_bool_and(x, y):
+    return bool_and(x, y)
+
+
+@infer(type=[(B, B, B),
+             (i64, B, InferenceError),
+             (B, i64, InferenceError)])
+def test_bool_or(x, y):
+    return bool_or(x, y)
+
+
+@infer(
+    type=[
+        (B, i64, i64, i64),
+        (i64, i64, i64, InferenceError),
+        (B, i64, f64, InferenceError),
+        ({'value': True}, i64, f64, i64),
+        ({'value': False}, i64, f64, f64),
```

Comment:
I'm not convinced that those two last cases should work.  We should spec switch so that types on both branches must match in all cases since they are both evaluated.

Reply:
Tricky. Consider a statement like `x = switch(hastype(x, Array), identity, to_array)(x)` which ensures that `x` is an `Array`. If `x` is `i64`, `to_array` has type signature `i64 -> Array(i64)`, but `identity` does not. If we force the evaluation of both branches, that statement will fail. Do we want that statement to fail? Possibly. But that removes the one use case I had for `switch` in the core.

Reply:
I guess that's fine if switch is only used as the backend for if (and maybe a few special uses in the core).

Reply:
I think it's fine from the user's perspective as well. Even if the condition is known to be True, the specializer will ask for the type of both branches nevertheless, and if evaluating the False branch would cause a type error, there will be an InferenceError. Basically, code that would cause an error in Python will also cause an error here.

Reply:
Ok then.